jwql/instrument_monitors/common_monitors/readnoise_monitor.py

```diff
@@ -0,0 +1,625 @@
+
+                    # Dont process files that already exist in the readnoise stats database
+                    file_exists = self.file_exists_in_database(output_filename)
+                    if file_exists:
+                        logging.info('\t{} already exists in the readnoise database table.'.format(output_filename))
+                        continue
+
+                    # Save any new uncal files with enough groups in the output directory; some dont exist in JWQL filesystem
+                    try:
+                        filename = filesystem_path(file_entry['filename'])
+                        uncal_filename = filename.replace('_dark', '_uncal')
+                        if not os.path.isfile(uncal_filename):
+                            logging.info('\t{} does not exist in JWQL filesystem, even though {} does'.format(uncal_filename, filename))
+                        else:
+                            n_groups = fits.getheader(uncal_filename)['NGROUPS']
+                            if n_groups > 1:  # skip processing if the file doesnt have enough groups to calculate the readnoise; TODO change to 10 before incorporating MIRI
+                                shutil.copy(uncal_filename, self.data_dir)
+                                logging.info('\tCopied {} to {}'.format(uncal_filename, output_filename))
+                                set_permissions(output_filename)
+                                new_files.append(output_filename)
+                            else:
+                                logging.info('\tNot enough groups to calculate readnoise in {}'.format(uncal_filename))
+                    except FileNotFoundError:
+                        logging.info('\t{} does not exist in JWQL filesystem'.format(file_entry['filename']))
+
+                # Run the readnoise monitor on any new files
+                if len(new_files) > 0:
+                    self.process(new_files)
+                    monitor_run = True
+                else:
+                    logging.info('\tReadnoise monitor skipped. {} new dark files for {}, {}.'.format(len(new_files), instrument, aperture))
+                    monitor_run = False
+
+                # Update the query history
+                new_entry = {'instrument': instrument,
+                             'aperture': aperture,
+                             'start_time_mjd': self.query_start,
+                             'end_time_mjd': self.query_end,
+                             'entries_found': len(new_entries),
+                             'files_found': len(new_files),
+                             'run_monitor': monitor_run,
+                             'entry_date': datetime.datetime.now()}
+                self.query_table.__table__.insert().execute(new_entry)
+                logging.info('\tUpdated the query history table')
+
+        logging.info('Readnoise Monitor completed successfully.')
+
+    def run_early_pipeline(self, filename, group_scale=False):
```

Comment:
I'm thinking it might be better to move this to the `pipeline_tools` module? In case it can eventually be used by other monitors.

Reply:
I changed to using the one that exists already in pipeline_tools - previously, I didn't use that one because I wanted to run custom settings for some steps (e.g. odd_even_rows=False for the refpix step), but I think this will be taken care of now that the pipeline is using these new parameter reference files to correctly set the defaults (i.e. NIRCam files will run the refpix step with that setting by default now)